internal/config/main/parse.go

```diff
@@ -0,0 +1,244 @@
+		{{- end }}
+
+	{{- end }}
+
+	{{- end }}
+}
+`
+)
+
+var (
+	fileTemplate = template.Must(template.New("gen").Funcs(sprig.TxtFuncMap()).Funcs(template.FuncMap{
+		"backtick": func() string {
+			return "`"
+		},
+	}).Parse(fileTemplateStr))
+)
+
+type flagType string
+
+const (
+	stringType      flagType = "string"
+	integerType     flagType = "integer"
+	booleanType     flagType = "boolean"
+	floatType       flagType = "float"
+	stringSliceType flagType = "stringSlice"
+	objectType      flagType = "object"
+)
+
+type flagDesc struct {
+	Name        string
+	Parent      string
+	Type        flagType
+	Description string
+	AddFlag     bool
+}
+
+type templateElem struct {
+	FlagDesc  flagDesc
+	ParamJSON string
+}
+
+func getName(member types.Member) string {
+	if jsonTag := reflect.StructTag(member.Tags).Get("json"); jsonTag != "" {
+		name, _ := stringutils.Split2(jsonTag, ",")
+		if name != "" {
+			return strcase.ToKebab(member.Name)
+		}
+	}
+	return strcase.ToKebab(member.Name)
+}
+
+func getDescription(member types.Member) string {
+	firstCommentLineWithMetadata := len(member.CommentLines)
+	for i, commentLine := range member.CommentLines {
+		if strings.HasPrefix(commentLine, metadataMarker) {
+			firstCommentLineWithMetadata = i
+			break
+		}
+	}
+	return strings.Join(member.CommentLines[:firstCommentLineWithMetadata], " ")
+}
+
+// hasViperExcludeTag checks if provided tags has viper=exclude tag
+func hasViperExcludeTag(tags map[string][]string) bool {
+	viperTags := tags[viperTagKey]
+	return sliceutil.Contains(viperTags, "exclude")
+
+}
+
+func constructFlagDescFromStruct(typeSpec *types.Type) ([]flagDesc, error) {
+	var flagDescs []flagDesc
+	for _, member := range types.FlattenMembers(typeSpec.Members) {
+
+		extractedTags := types.ExtractCommentTags("+", member.CommentLines)
+		if hasViperExcludeTag(extractedTags) {
+			continue
+		}
+
+		desc := flagDesc{
+			Name:        getName(member),
+			Description: getDescription(member),
+			AddFlag:     true,
+		}
+
+		relevantTyp := member.Type
+
+		switch kind := relevantTyp.Kind; kind {
+		case types.Builtin:
+			switch relevantTyp {
+			case types.String:
+				desc.Type = stringType
+			case types.Int:
+				desc.Type = integerType
+			case types.Float32, types.Float64:
+				desc.Type = floatType
+			case types.Bool:
+				desc.Type = booleanType
+			default:
+				return nil, errors.Errorf("currently unsupported type %v", member.Type)
+			}
+		case types.Struct:
+			desc.Type = objectType
+			desc.AddFlag = false
+			subDesc, err := constructFlagDescFromStruct(member.Type)
+			// Template uses Parent name in mapstructure in order to
+			// correctly map config values
+			for i := range subDesc {
+				subDesc[i].Parent = desc.Name
+			}
+
+			if err != nil {
+				return nil, errors.Wrapf(err, "handling field %v", member.Name)
+			}
+			flagDescs = append(flagDescs, subDesc...)
+		case types.Slice:
```

Comment:
could you please suggest a better way to handle `[]string`?

Reply:
We can just make sure that the underlying type is a string, and return an error if it's not. That way, if someone adds a different kind of parameter, they know they will need to update this code.